### PR TITLE
Add native implementation for soft nms

### PR DESCRIPTION
**Pull request by willthefrog**
and refactor a bit

experiments show same or improved mAP.

### CODE REVIEW

File: src/nms.cc [hunk 1]
```diff
@@ -0,0 +1,86 @@
+#include <cmath>
+#include <algorithm>
+
+namespace nms
+{
```
Comment (qingqing01) [inline src/nms.cc]: code style in Paddle repo is Google C++ code style:

```
namespace nms {

}
```

“The open curly brace is always on the end of the last line of the function declaration, not the start of the next line.”

https://google.github.io/styleguide/cppguide.html#Function_Declarations_and_Definitions

File: src/binding.cc [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  m.doc() = "SoftNMS for object detection.";
+
+  py::enum_<nms::Method>(m, "NMSMethod")
+    .value("LINEAR", nms::Method::LINEAR)
+    .value("GAUSSIAN", nms::Method::GAUSSIAN)
+    .value("HARD", nms::Method::HARD)
+    .export_values();
+  m.def("soft_nms", &py_soft_nms, "boxes"_a.noconvert(),
+        "method"_a = nms::Method::GAUSSIAN,
+        "Nt"_a = 0.3, "sigma"_a = 0.5, "threshold"_a = 0.001);
```
[the remark this answers is not in the document]
Comment (willthefrog) [reply src/binding.cc]: pybind11 literal

File: ppdet/modeling/ops.py [hunk 3]
```diff
@@ -216,102 +217,122 @@ def __init__(self,
                  softnms_sigma=0.5,
                  normalized=False,
                  background_label=0,
-                ):
+                 Nt=0.3,
+                 method='gaussian',
+                 use_cpp=False):
```
[the remark this answers is not in the document]
Comment (willthefrog) [reply ppdet/modeling/ops.py]: done

File: src/nms.cc (hunk 1 above)
Comment (willthefrog) [reply src/nms.cc]: done

**Pull request closed without merging**